pkg/daemon/daemon.go

```diff
@@ -417,49 +449,65 @@ func (dn *Daemon) CheckStateOnBoot() error {
 		select {}
 	}
 
-	inDesiredConfig := stateAndConfigs.currentConfig == stateAndConfigs.desiredConfig
-	var completingPendingUpdate bool
-	if inDesiredConfig {
-		glog.Infof("In desired config %s", stateAndConfigs.currentConfig.GetName())
-		// The MCS used to not write the state=done annotation key, let's do that here for
-		// now too.
-		if stateAndConfigs.state == "" {
-			if err := dn.nodeWriter.SetUpdateDone(dn.kubeClient.CoreV1().Nodes(), dn.name, stateAndConfigs.desiredConfig.GetName()); err != nil {
-				return err
-			}
-		}
-		completingPendingUpdate = false
+	// Validate the on-disk state against what we *expect*.
+	//
+	// In the case where we're booting a node for the first time, or the MCD
+	// is restarted, that will be the current config.
```

Comment:
My worry with this is that we expect the MCD to always be restarted when the machine state is clearly defined. E.g. what if the MCD gets OOM-killed while writing out files? Can we relax this a bit in the following way?

- in the pendingConfig case, error out if validation fails
- in the case currentConfig == desiredConfig already, error out if validation fails
- otherwise, just try to apply the desiredConfig

Reply:
Well...we could introduce a "writingConfig".  

I dunno though...I see your point, but I feel like the best course is to land this and iterate from there.  Having the MCD die half-way through is a problem we could solve better for sure, but...worst case scenario is that validation fails, we go degraded, and the machine eventually gets reprovisioned right?  Wwe also have bigger problems like needing the retry loops around fetching configs that are impacting us a lot more today.

Reply:
I'm mentioning it because this is behaviour we're regressing on, right? I.e. the current code does handle this case. Not sure in prod clusters how likely it is, but it's definitely helpful for when iterating on the MCD itself on the bits that actually perform the update. I'm good with doing this as a follow-up though.